Enhancement in Ncat - Broadcast data by using delimiter

rewanth1997 opened the pull request:
**Enhancement in Ncat - Broadcast data by using delimiter**

`--use-delimiter` option added to Ncat. It can be used while broadcasting the data. Use delay option to have see the change in how the data is being sent on using delimiter.

**Note :** If delay option is not used you cannot observe any change. The whole file will be broadcasted in a single instance.

`--use-delimiter` expects the ascii value of the delimiter as an argument.

0, -1 are assumed to be the ASCII values for EOF. If these are used the whole file will be broadcasted as a whole.

Change to the Ncat directory in terminal.
*Terminal 1* : `cat INSTALL | ./ncat --use-delimiter 10 -d 2 -l 1337 -vv`
*Terminal 2* : `./ncat localhost 1337`

**Explanation of the above commands used in Terminal 1**
* `--use-delimiter` takes 10 as an argument which is the ascii value of next line.
* `-d` takes 2 as an argument and it delays broadcasting data by 2 seconds.
   *   It sleeps for every 2 seconds while broadcasting data, when it hits the delimiter.
* `-l` takes 1337 as an argument which is the port number to be listened on.
*`-vv` for verbose output, using this will let you see when the file reaches EOF.

**Output**
The data is broadcasted in chunks when delimiter is met in the buffer. The program sleeps for every two seconds after broadcasting the data.

To have more closer observation, you can pass `101` as an argument to `--use-delimiter` which breaks the data till `e` everytime before broadcasting.

Terminal 1 and Terminal 2 are using `e` as a delimiter (ASCII value of `e` is 101)
*Terminal 1* : `cat INSTALL | ./ncat --use-delimiter 101 -d 1 -l 1337 -vv`
*Terminal 2* : `./ncat localhost 1337`

Terminal 3 and Terminal 4 are using EOF as a delimiter (ASCII value of EOF taken as 0), the complete file will be sent in a single shot.
*Terminal 3* : `cat INSTALL | ./ncat --use-delimiter 0 -d 1 -l 1337 -vv`
*Terminal 4* : `./ncat localhost 1337`

AnshulMalik commented:
Tamannah, Is it really you ?

rewanth1997 commented:
Yes, @AnshulMalik its me.

dmiller-nmap commented on ncat/ncat_core.c:
```diff
@@ -60,7 +60,7 @@
  * listed in the included docs/licenses/OpenSSL.txt file, and distribute   *
  * linked combinations including the two.                                  *
- *                                                                         * 
+ *                                                                         *
```
These whitespace changes will be fixed if you rebase the patch; this might require some extra work, because I couldn't get the patch to apply cleanly, but it's something you'd have to do anyway before merging in svn.

dmiller-nmap commented on ncat/ncat_core.c:
```diff
@@ -171,6 +171,8 @@ void options_init(void)
     o.broker = 0;
     o.listen = 0;
+    o.delimiter = 0;
+    o.delimiter_used = 0; /* This gets activated when delimiter option gets enabled. (Works as a flag) */
```
Since you're using an `int` for `delimiter` but only using one byte of it (casting to `char`), you can use the sign to represent this flag instead. If `o.delimiter` is less than 0, it is not set. Using a whole `int` is a little wasteful; you could use a `short` instead, since you're only using the sign and one byte.

dmiller-nmap commented on ncat/ncat_listen.c:
```diff
@@ -621,6 +627,65 @@ int read_stdin(void)
+    char buf[DEFAULT_TCP_BUF_LEN];
+
+    /* Converting the ascii valued delimiter parameter to character. */
+    char delimiter = (int)o.delimiter;
```
This casts an `int` to an `int`. I think you meant to cast to a `char`. But the function would be nicer if `delimiter` was an argument. Make the caller do the casting.

dmiller-nmap commented on ncat/ncat_listen.c:
```diff
@@ -621,6 +627,65 @@ int read_stdin(void)
+    /* Converting the ascii valued delimiter parameter to character. */
+    char delimiter = (int)o.delimiter;
+
+    char tc[1];       /* Temporary character */
```
You are reading 1 byte at a time, which will kill performance. Better to do some sort of buffered approach where you read in a good amount then split it on the delimiter and send; if there is leftover, move it to the front of the buffer and read again up to the size of the buffer.

dmiller-nmap commented on ncat/ncat_listen.c:
```diff
@@ -621,6 +627,65 @@ int read_stdin(void)
+    do {
+        nbytes = read(STDIN_FILENO, tc, sizeof(tc));
+        buf[index++] = tc[0];
+    } while(index < DEFAULT_TCP_BUF_LEN && tc[0] != delimiter && tc[0] != '\0' && nbytes != 0);
```
Why the `tc[0] != '\0'` check? This prevents sending data containing null bytes, which does occur.
Also, need to check for `nbytes < 0`, which indicates an error.

dmiller-nmap commented on ncat/ncat_listen.c:
```diff
@@ -621,6 +627,65 @@ int read_stdin(void)
+
+        /* The data in the buffer along with the EOF added to the broadcast set. */
+        if(index > 1)
+          ncat_broadcast(&master_broadcastfds, &broadcast_fdlist, buf, index - 1);
```
This doesn't belong here; the call to ncat_broadcast is farther down.

dmiller-nmap commented on ncat/ncat_listen.c:
```diff
@@ -621,6 +627,65 @@ int read_stdin(void)
+        buf[index++] = tc[0];
+    } while(index < DEFAULT_TCP_BUF_LEN && tc[0] != delimiter && tc[0] != '\0' && nbytes != 0);
+
+    if (nbytes <= 0) {
```
Apart from the lines 664-666, which I argue should not be here, the rest of this function is identical to `read_stdin`. Put the differing code into a smaller function and have `read_stdin` call that if `o.delimiter` is set.

dmiller-nmap commented on ncat/ncat_main.c:
```diff
@@ -280,6 +280,7 @@ int main(int argc, char *argv[])
         {"lua-exec",        required_argument,  NULL,         0},
         {"lua-exec-internal",required_argument, NULL,         0},
 #endif
+        {"use-delimiter",   required_argument,  NULL,         'b'},
```
Please report to me whether traditional netcat or openbsd netcat uses the `-b` short option and what they use it for. This seems to be a new functionality that is not part of either one, so it does not need a short option.

dmiller-nmap commented on ncat/ncat_main.c:
```diff
@@ -618,6 +626,7 @@ int main(int argc, char *argv[])
 "  -w, --wait <time>          Connect timeout\n"
 "  -z                         Zero-I/O mode, report connection status only\n"
 "      --append-output        Append rather than clobber specified output files\n"
+"      --use-delimiter        Breaks the data till delimiter before broadcasting data; expects ascii value of character (Max range: 255)\n"
```
The option name should be `--delimiter`. `--use-delimiter` sounds more like an on/off flag, and this requires an argument. Be sure to note the argument in the help text as `<byte value>`. Description should not mention "broadcast" because that means something different to the user ("send to a broadcast IP address") than it does in our code ("write to all the appropriate file descriptors"). Value is not "ascii" but the decimal byte value 0-255.

dmiller-nmap commented on ncat/ncat_main.c:
```diff
@@ -652,6 +661,12 @@ int main(int argc, char *argv[])
+    /*  --use-delimiter can be used only for broadcasting purposes.
+        So checking for listen mode is necessary before moving forward. */
+    if(o.delimiter_used == 1 && o.listen != 1) {
+        bye("--use-delimiter option is binded with -l option.");
```
This error message is hard to understand. Copy a different one and just change the option name. But our goal should be to use this any time we are reading stdin, which would include listen mode as well. Look into it, because that would be the next project we'd tackle if we get this one done.

rewanth1997 commented:
The new function makes request in a different way.
Delimiter argument is taken as string instead of decimal numbers.

*Terminal 1:*
```
cat INSTALL | ./ncat --delimiter="e" -l 1337 -vv -d 0.5
```
*Terminal 2:*
```
./ncat localhost 1337
```

rewanth1997 commented:
I removed the old code and added a new function which handles the data in entirely different way. Please ignore some unwanted commits above and review the new code directly.
